### PR TITLE
Added EB rule for ip addresses in conflist for linux

### PR DESCRIPTION
**Pull request by pjohnst5**
Added EB rules for ip addresses in the conflist. These rules route these ip addresses via the host for linux.

### CODE REVIEW

File: network/endpoint.go [hunk 1]
```diff
@@ -55,6 +56,7 @@ type EndpointInfo struct {
 	MacAddress               net.HardwareAddr
 	DNS                      DNSInfo
 	IPAddresses              []net.IPNet
+	IPsToRouteViaHost        []string
```
Comment (pjohnst5) [inline network/endpoint.go]: @tamilmani1989 If I remove this line, make all-binaries fails saying:

> epInfo.IPsToRouteViaHost undefined (type *EndpointInfo has no field or method IPsToRouteViaHost)

File: network/endpoint.go [hunk 2]
```diff
@@ -25,6 +25,7 @@ type endpoint struct {
 	MacAddress               net.HardwareAddr
 	InfraVnetIP              net.IPNet
 	LocalIP                  string
+	IPsToRouteViaHost        []string
```
Comment (pjohnst5) [inline network/endpoint.go]: @tamilmani1989 This line throws no errors when removed so I can remove it

File: network/endpoint.go (hunk 1 above)
Comment (tamilmani1989) [reply network/endpoint.go]: dont remove it from endpointinfo..just remove from endpoint

File: ebtables/ebtables.go [hunk 3]
```diff
@@ -51,6 +52,22 @@ func SetArpReply(ipAddress net.IP, macAddress net.HardwareAddr, action string) e
 	return executeShellCommand(command)
 }
 
+// SetEBRule sets an EB rule to route the given IP via the host.
+func SetEBRule(ipAddress string) error {
```
Comment (tamilmani1989) [inline ebtables/ebtables.go]: rename this function to SetBrouteAccept and it should take 2 parmeters as like other function, 1 is ipaddress and other is action -> Either add or delete

File: ebtables/ebtables.go [hunk 4]
```diff
@@ -51,6 +52,22 @@ func SetArpReply(ipAddress net.IP, macAddress net.HardwareAddr, action string) e
...
+	command := fmt.Sprintf(
+		"ebtables -t broute -A BROUTING --ip-dst %s -p IPv4 -j redirect --redirect-target ACCEPT",
+		ipAddress)
+
+	return executeShellCommand(command)
+}
+
+// GetEBRules gets the EB rules to route via host.
+func GetEBRules() (string, error) {
```
Comment (tamilmani1989) [inline ebtables/ebtables.go]: rename this to GetEbtableRules(tableName, chainName string)

here tableName is broute and chainName is BROUTING

File: network/bridge_endpointclient_linux.go [hunk 5]
```diff
@@ -81,6 +83,26 @@ func (client *LinuxBridgeEndpointClient) AddEndpointRules(epInfo *EndpointInfo)
...
+	for _, ipAddr := range epInfo.IPsToRouteViaHost {
+		// Get EB rules.
+		log.Printf("[net] Getting EB rules for routing via host")
+		out, err := ebtables.GetEBRules()
+		if err != nil {
+			log.Printf("[net] Failed to get EB rules: %v", err)
+			return err
+		}
+
+		log.Printf("[net] Checking if EB rule already exists for IP address %v", ipAddr)
```
Comment (tamilmani1989) [inline network/bridge_endpointclient_linux.go]: move this to generic function in ebtable package. can be

EbtableRuleExists(tableName, chainName, matchSet string). You can call getebtablerules function inside this call

File: ebtables/ebtables.go [hunk 6]
```diff
@@ -51,6 +52,61 @@ func SetArpReply(ipAddress net.IP, macAddress net.HardwareAddr, action string) e
...
+		tableName, chainName)
+	out, err := platform.ExecuteCommand(command)
+	if err != nil {
+		return nil, err
+	}
+
+	// Splits lines and finds rules.
+	lines := strings.Split(out, "\n")
+	chainTitle := fmt.Sprintf("Bridge chain: %s", chainName)
+	var inChain bool
```
Comment (tamilmani1989) [inline ebtables/ebtables.go]: leave blank.
 also move this var to top

var (
 inChain bool
rules []string
)
)

File: ebtables/ebtables.go [hunk 7]
```diff
@@ -51,6 +52,61 @@ func SetArpReply(ipAddress net.IP, macAddress net.HardwareAddr, action string) e
...
+func SetBrouteAccept(ipAddress, action string) error {
+	command := fmt.Sprintf(
+		"ebtables -t broute %s BROUTING --ip-dst %s -p IPv4 -j redirect --redirect-target ACCEPT",
+		action, ipAddress)
+
+	return executeShellCommand(command)
+}
+
+// GetEbtableRules gets EB rules for a table and chain.
+func GetEbtableRules(tableName, chainName string) ([] string, error) {
```
Comment (tamilmani1989) [inline ebtables/ebtables.go]: remove space between [] and string

File: ebtables/ebtables.go [hunk 8]
```diff
@@ -51,6 +52,61 @@ func SetArpReply(ipAddress net.IP, macAddress net.HardwareAddr, action string) e
...
+		}
+		if inChain {
+			if strings.HasPrefix(line, "-") {
+				rules = append(rules, strings.TrimSpace(line))
+			} else {
+				break
+			}
+		}
+	}
+	return rules, nil
```
Comment (tamilmani1989) [inline ebtables/ebtables.go]: leave a blank line before return

File: ebtables/ebtables.go [hunk 9]
```diff
@@ -51,6 +52,61 @@ func SetArpReply(ipAddress net.IP, macAddress net.HardwareAddr, action string) e
...
+	if err != nil {
+		return false, err
+	}
+
+	for _, rule := range rules {
+		if rule == matchSet {
+			return true, nil
+		}
+	}
+	return false, nil
```
Comment (tamilmani1989) [inline ebtables/ebtables.go]: blank line before return

File: ebtables/ebtables.go [hunk 10]
```diff
@@ -51,6 +52,67 @@ func SetArpReply(ipAddress net.IP, macAddress net.HardwareAddr, action string) e
 	return executeShellCommand(command)
 }
 
+// SetBrouteAccept sets an EB rule.
+func SetBrouteAccept(ipAddress, action string) (string, error) {
+	command := fmt.Sprintf(
+		"ebtables -t broute %s BROUTING --ip-dst %s -p IPv4 -j redirect --redirect-target ACCEPT",
+		action, ipAddress)
+
+	return platform.ExecuteCommand(command)
```
Comment (tamilmani1989) [inline ebtables/ebtables.go]: just return error thats fine

_, err := platform.ExecuteCommand(command)
return err

File: network/bridge_endpointclient_linux.go [hunk 11]
```diff
@@ -81,6 +82,32 @@ func (client *LinuxBridgeEndpointClient) AddEndpointRules(epInfo *EndpointInfo)
 		}
 	}
 
+	for _, ipAddr := range epInfo.IPsToRouteViaHost {
```
Comment (tamilmani1989) [inline network/bridge_endpointclient_linux.go]: add this as separate funciton

File: network/bridge_endpointclient_linux.go [hunk 12]
```diff
@@ -174,3 +177,33 @@ func (client *LinuxBridgeEndpointClient) DeleteEndpoints(ep *endpoint) error {
 
 	return nil
 }
+
+func AddRuleToRouteViaHost(epInfo *EndpointInfo) error {
```
Comment (tamilmani1989) [inline network/bridge_endpointclient_linux.go]: AddRuleToRouteViaHost- make this to addRuleToRouteViaHost
The reason is in golang, if function/variable start with caps, its like exported function/variable.

**Pull request merged**